course_discovery/apps/edx_elasticsearch_dsl_extensions/management/commands/update_index.py

```diff
@@ -0,0 +1,219 @@
+        conn = get_connection()
+        if not record_count_is_sane:
+            attempts = 0
+            while attempts < 2:
+                attempts += 1
+                current_attempt_record_count = self.get_record_count(document)
+                current_attempt_percentage_change = self.percentage_change(
+                    current_attempt_record_count, previous_record_count)
+                alternate_current_record_count = conn.search({"query": {"match_all": {}}}, index=new_index_name).get(
+                    'hits', {}).get('total', {}).get('value', 0)
+                message = '''
+        Sanity check failed for attempt #{0}.
+        Index name: {1}
+        Percentage change: {2}
+        Base record count: {3}
+        Search record count: {4}
+                    '''.format(
+                    attempts,
+                    new_index_name,
+                    str(int(round(current_attempt_percentage_change * 100, 0))) + '%',
+                    current_attempt_record_count,
+                    alternate_current_record_count
+                )
+                logger.info(message)
+                logger.info('...sleeping for 5 seconds...')
+                time.sleep(5)
+
+        index_info_string = (
+            'The previous index contained [{}] records. '
+            'The new index contains [{}] records, a [{:.2f}%] change.'.format(
+                previous_record_count, current_record_count, percentage_change * 100
+            )
+        )
+
+        return record_count_is_sane, index_info_string
+
+    @staticmethod
+    def get_record_count(document):
+        return document.search().query().count()
+
+    @contextlib.contextmanager
+    def preserve_state_registered_indexes(self, state_to_update, models):
+        # pylint: disable=protected-access
+        for index, (__, new_name) in zip(registry.get_indices(models), state_to_update):
+            index._name = new_name
+        yield
+        for index, (existed_name, __) in zip(registry.get_indices(models), state_to_update):
+            index._name = existed_name
+
+    def prepare_backend_index(self, registered_index, backend='default'):
+        """
+        Prepares an index that will be used to store data by the backend.
+
+        Args:
+            registered_index (Index): django elasticsearch index instance.
+            backend (ElasticsearchSearchBackend): Backend to update.
+
+        Returns:
+            (tuple): tuple containing:
+
+                alias(str): Recommended alias for the new index.
+                index_name(str): Name of the newly-created index.
+        """
+        copied_registered_index = copy(registered_index)
+        timestamp = datetime.datetime.utcnow().strftime('%Y%m%d_%H%M%S')
+        # pylint: disable=protected-access
+        existed_index_name = copied_registered_index._name
+        alias, *_ = copied_registered_index.get_alias(using=backend).get(existed_index_name, {}).get('aliases').keys()
```

Comment:
This line failed for me locally. 

I had first done a `./manage.py search_index --create` to create the indices with aliases.
Then ran `update_index` and got `AttributeError: 'NoneType' object has no attribute 'keys'` on this line.

Doing some inspection, the `get_alias` call is returning something like `{'course_20200825_165521': {'aliases': {'course': {}}}}` but the incoming `existed_index_name` is the alias of `course` -- looks like the surrounding call expects it to be the actual index name of `course_xxx_yyy`. The toplevel `registry.get_indices(models)` call is returning indices with the aliases as the `_name` attributes.

Reply:
Thanks, I redid this part to simplify and fix. Pushed new commits. Please fetch and do:
1. ./manage.py search_index --create
2. ./manage.py search_index --populate
3. ./manage.py update_index

Please post your feedback after I want it to analyze and understand where we are. Thanks in advance.